avocado_vt/plugins/vt_cluster.py

```diff
@@ -0,0 +1,246 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2025
+# Authors: Yongxue Hong <yhong@redhat.com>
+
+"""
+This module provides an Avocado-VT plugin for managing a test cluster.
+
+The `VTCluster` plugin hooks into the job lifecycle to perform setup before
+tests run and cleanup after they complete. It is responsible for starting and
+stopping agent servers on cluster nodes, handling metadata, and collecting
+logs. This ensures that the test environment is properly configured and torn
+down for cluster-based tests.
+"""
+import logging
+import os
+import sys
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+from avocado.core import exit_codes
+from avocado.core.plugin_interfaces import JobPostTests as Post
+from avocado.core.plugin_interfaces import JobPreTests as Pre
+from avocado.utils.stacktrace import log_exc_info
+
+from virttest.vt_cluster import cluster, node_metadata
+
+
+class ClusterSetupError(Exception):
+    """
+    Represents any error situation when attempting to create a cluster.
+    """
+
+    pass
+
+
+class ClusterManagerSetupError(ClusterSetupError):
+    """
+    Represents an error during the setup of the cluster manager.
+    """
+
+    pass
+
+
+class ClusterCleanupError(Exception):
+    """
+    Represents an error during the cleanup of the cluster.
+    """
+
+    pass
+
+
+class ClusterManagerCleanupError(ClusterCleanupError):
+    """
+    Represents an error during the cleanup of the cluster manager.
+    """
+
+    pass
+
+
+class VTCluster(Pre, Post):
+    """
+    An Avocado-VT plugin to set up and tear down a multi-node cluster.
+
+    This plugin manages the lifecycle of a test cluster by implementing the
+    `pre_tests` and `post_tests` hooks. It ensures that cluster nodes are
+    initialized with agent servers before tests begin and are properly shut
+    down and cleaned up after the job finishes.
+    """
+
+    name = "vt-cluster"
+    description = "Avocado-VT Cluster Pre/Post"
+
+    def __init__(self, **kwargs):
+        """
+        Initializes the VTCluster plugin.
+        """
+        self._log = logging.getLogger("avocado.app")
+
+    @staticmethod
+    def _setup_node():
+        """
+        Starts agent servers on all cluster nodes in parallel.
+
+        :raise ClusterSetupError: If starting an agent fails.
+        """
+        nodes = cluster.get_all_nodes()
+        if not nodes:
+            return
+
+        def __start_node_agent(node):
+            """Helper function to start agent on a single node."""
+            try:
+                node.start_agent_server()
+                return node, None
+            except Exception as err:
+                return node, err
+
+        try:
+            with ThreadPoolExecutor(max_workers=len(nodes)) as executor:
```

Comment:
is master node included? It looks cluster.get_all_nodes() will return both master node and worker nodes.